File: kt/1465.kt
const val MOD = 1000000007
class Solution {
    fun maxArea(h: Int, w: Int, horizontalCuts: IntArray, verticalCuts: IntArray): Int {
        horizontalCuts.sort()
        verticalCuts.sort()
        return (maxLength(horizontalCuts, h).toLong() * maxLength(verticalCuts, w).toLong() % MOD ).toInt()
    }

    fun maxLength(arr: IntArray, boardLength: Int): Int {
        var pre = 0
        var max = 0
        for (i in arr.indices) {
            val cur = arr[i]
            max = max.coerceAtLeast(cur - pre)
            pre = cur
        }

        return max.coerceAtLeast(boardLength - pre)
    }
}

File: ts/1465.ts
function maxArea(
    h: number,
    w: number,
    horizontalCuts: number[],
    verticalCuts: number[]
): number {
    horizontalCuts.sort((a, b) => a - b)
    verticalCuts.sort((a, b) => a - b)
    return Number(
        (BigInt(findMax(horizontalCuts, h)) *
            BigInt(findMax(verticalCuts, w))) %
            BigInt(1000000007)
    )
}

const findMax = (arr: number[], board: number): number => {
    let prev = 0,
        max = 0
    for (let cur of arr) {
        max = Math.max(max, cur - prev)
        prev = cur
    }

    return Math.max(max, board - prev)
}
